ldelf/ta_elf.c

```diff
@@ -0,0 +1,811 @@
+			p += 3;
+		else
+			p += 2;
+	}
+	*uuid = u;
+out:
+	return res;
+}
+
+static void add_deps_from_segment(struct ta_elf *elf, unsigned int type,
+				  vaddr_t addr, size_t memsz)
+{
+	size_t dyn_entsize = 0;
+	size_t num_dyns = 0;
+	size_t n = 0;
+	unsigned int tag = 0;
+	size_t val = 0;
+	TEE_UUID uuid = { };
+	char *str_tab = NULL;
+
+	if (type != PT_DYNAMIC)
+		return;
+
+	if (elf->is_32bit)
+		dyn_entsize = sizeof(Elf32_Dyn);
+	else
+		dyn_entsize = sizeof(Elf64_Dyn);
+
+	assert(!(memsz % dyn_entsize));
+	num_dyns = memsz / dyn_entsize;
+
+	for (n = 0; n < num_dyns; n++) {
+		read_dyn(elf, addr, n, &tag, &val);
+		if (tag == DT_STRTAB) {
+			str_tab = (char *)(val + elf->load_addr);
+			break;
+		}
+	}
+
+	for (n = 0; n < num_dyns; n++) {
+		read_dyn(elf, addr, n, &tag, &val);
+		if (tag != DT_NEEDED)
+			continue;
+		parse_uuid(str_tab + val, &uuid);
+		queue_elf(&uuid);
+	}
+}
+
+static void add_dependencies(struct ta_elf *elf)
+{
+	size_t n = 0;
+
+	if (elf->is_32bit) {
+		Elf32_Phdr *phdr = elf->phdr;
+
+		for (n = 0; n < elf->e_phnum; n++)
+			add_deps_from_segment(elf, phdr[n].p_type,
+					      phdr[n].p_vaddr, phdr[n].p_memsz);
+	} else {
+		Elf64_Phdr *phdr = elf->phdr;
+
+		for (n = 0; n < elf->e_phnum; n++)
+			add_deps_from_segment(elf, phdr[n].p_type,
+					      phdr[n].p_vaddr, phdr[n].p_memsz);
+	}
+}
+
+static void copy_section_headers(struct ta_elf *elf)
+{
+	TEE_Result res = TEE_SUCCESS;
+	size_t sz = elf->e_shnum * elf->e_shentsize;
+	size_t offs = 0;
+
+	elf->shdr = malloc(sz);
+	if (!elf->shdr)
+		err(TEE_ERROR_OUT_OF_MEMORY, "malloc");
+
+	/*
+	 * We're assuming that section headers comes after the load segments,
+	 * but if it's a very small dynamically linked library the section
+	 * headers can still end up (partially?) in the first mapped page.
+	 */
+	if (elf->e_shoff < SMALL_PAGE_SIZE) {
+		assert(!elf->is_main);
+		offs = MIN(SMALL_PAGE_SIZE - elf->e_shoff, sz);
+		memcpy(elf->shdr, (void *)(elf->load_addr + elf->e_shoff),
+		       offs);
+	}
+
+	if (offs < sz) {
+		res = sys_copy_from_ta_bin((uint8_t *)elf->shdr + offs,
+					   sz - offs, elf->handle,
+					   elf->e_shoff + offs);
+		if (res)
+			err(res, "sys_copy_from_ta_bin");
+	}
+}
+
+static void close_handle(struct ta_elf *elf)
+{
+	TEE_Result res = sys_close_ta_bin(elf->handle);
+
+	if (res)
+		err(res, "sys_close_ta_bin");
+	elf->handle = -1;
+}
+
+void ta_elf_load_main(const TEE_UUID *uuid, uint32_t *is_32bit,
+		      uint64_t *entry, uint64_t *sp, uint32_t *ta_flags)
+{
+	struct ta_elf *elf = queue_elf(uuid);
+	struct ta_head *head;
+	vaddr_t va = 0;
+	TEE_Result res = TEE_SUCCESS;
+
+	assert(elf);
+	elf->is_main = true;
+
+	init_elf(elf);
+
+	/*
+	 * Legacy TAs doesn't set entry point, instead it's set in ta_head.
+	 * If entry point isn't set explicitly, set to the start of the
+	 * first executable section by the linker. Since ta_head also
+	 * always comes first in legacy TA it means that the entry point
+	 * will be set to 0x20.
```

Comment:
Perhaps mention when the transition from "legacy" occurred? (although it is not that hard to find).
=> "`Since commit a73b5878c89d ("Replace ta_head.entry with elf entry")`".

Reply:
I'll fix